core/src/main/scala/org/ensime/indexer/SearchService.scala

```diff
@@ -316,50 +383,53 @@ class IndexingQueueActor(searchService: SearchService) extends Actor with ActorL
 
   override def receive: Receive = {
     case IndexFile(f) =>
-      todo += f.getName.getURI -> f
+      f match {
+        case jar if jar.getName.getExtension == "jar" => todo.addBinding(jar, jar)
+        case classFile => todo.addBinding(searchService.getTopLevelClassFile(classFile), classFile)
+      }
       debounce()
 
     case Process if todo.isEmpty => // nothing to do
 
     case Process =>
-      val (batch, remaining) = todo.splitAt(500)
-      todo = remaining
-      if (remaining.nonEmpty)
+      val batch = todo.take(250)
+      batch.keys.foreach(todo.remove)
+      if (todo.nonEmpty)
         debounce()
 
       import ExecutionContext.Implicits.global
 
-      log.debug(s"Indexing ${batch.size} files")
+      log.debug(s"Indexing ${batch.size} groups of files")
 
       def retry(): Unit = {
-        batch.foreach(self !)
+        batch.valuesIterator.foreach(_.foreach(self ! IndexFile(_)))
       }
 
-      Future.sequence(batch.map {
-        case (url, f) =>
-          val filename = f.getName.getPath
+      batch.grouped(10).foreach(chunk => Future.sequence(chunk.map {
+        case (outerClassFile, _) =>
+          val filename = outerClassFile.getName.getPath
           // I don't trust VFS's f.exists()
           if (!File(filename).exists()) {
             Future {
               searchService.semaphore.acquire() // nasty, but otherwise we leak
-              f -> Nil
+              outerClassFile -> Nil
             }
-          } else searchService.extractSymbolsFromClassOrJar(f).map(f -> )
+          } else searchService.extractSymbolsFromClassOrJar(outerClassFile, batch).map(outerClassFile -> )
       }).onComplete {
         case Failure(t) =>
           searchService.semaphore.release()
           log.error(t, s"failed to index batch of ${batch.size} files. $advice")
           retry()
         case Success(indexed) =>
-          searchService.delete(indexed.map(_._1)(collection.breakOut)).onComplete {
+          searchService.delete(indexed.flatMap(f => batch(f._1))(collection.breakOut)).onComplete {
             case Failure(t) =>
               searchService.semaphore.release()
               log.error(t, s"failed to remove stale entries in ${batch.size} files. $advice")
               retry()
             case Success(_) => indexed.foreach {
               case (file, syms) =>
                 val boost = searchService.isUserFile(file.getName)
-                val persisting = searchService.persist(FileCheck(file), syms, commitIndex = true, boost = boost)
+                val persisting = searchService.persist(syms, commitIndex = true, boost = boost)
 
                 persisting.onComplete {
                   case _ => searchService.semaphore.release()
```

Comment:
it would be good to have some kind of feedback of when `.jar` files have been persisted... I can use it to report back how long it takes to re-index big jars at my work.